src/server/services/wms/qgswmsrenderer.cpp

```diff
@@ -3216,15 +3216,20 @@ namespace QgsWms
   {
     if ( layer->type() == QgsMapLayerType::VectorLayer )
     {
-      QgsFeatureIds selectedIds;
+      QgsVectorLayer *vl = qobject_cast<QgsVectorLayer *>( layer );
+
+      QgsFeatureRequest request;
+      QgsServerFeatureId::updateFeatureRequestFromServerFids( request, fids, vl->dataProvider() )
+      QgsFeatureIds selectedIds = request.filterFids();
 
-      for ( const QString &id : fids )
+      if ( selectedIds.empty() )
```

Comment:
Is this check meant for speed?
I think that the `else` here will work in all cases (when the pk is a single value and when it is a composite).

Reply:
In the `QgsServerFeatureId::updateFeatureRequestFromServerFids`, if the primary key attributes list is empty the filter fids is set else the expression is build. So the `else` is not available for all cases.

I like to fix the selection method, because we have an issue with selection in the case of a filtered postgis layer by access control plugin.